pkg/roachpb/api.proto

```diff
@@ -1010,6 +1010,18 @@ message GCRequest {
   // the request will fail.
   GCClearRangeKey clear_range_key = 7;
```

Comment:
My first take was to just add timestamp to existing clear range and then use timestamp to determine what to do with it. If no timestamp is specified and requested start/end key match range boundaries then proceed with fast path. If there's a timestamp or boundaries are within the range, then proceed with deletion of of sub range.
Those requests have slightly different semantics subrange only clears points, while full range del removes everything because it is triggered by schema change. If we mistake full range as subrange then range tombstones would be kept. But it is probably not bad as it could only happen on merge and we will collect them on next iteration.
For compatibility purposes, we would always have full range without timestamp which we would handle, and we would version gate requests to ensure we don't accidentally clean anything + we still do a live version check.
Does that approach sound reasonable to you?

Reply:
>subrange only clears points

Why can't we clear both? If the start key is versioned just use `startKey.Next()` as the start bound for the range key clear.

Reply:
I'm thinking of the case where you have a range tombstone over the range. If existing data will start falling out of visibility gradually, you will start clearing holes in this single tombstone as parts of the range become fully collectable. That would start multiplying number of range keys. And GC test show that if range is deleted, there would be few GC rounds before all the data before range tombstone is removed.
If we treat points and ranges separately, then we avoid this complexity and overhead.

Reply:
> If existing data will start falling out of visibility gradually, you will start clearing holes in this single tombstone as parts of the range become fully collectable.

We can't clear parts of that range tombstone using a clearrange until the range tombstone itself falls below the GC threshold. When it does, we can clear the whole thing. If there's any visible data below the range tombstone, then we can't GC that either until the range tombstone falls below the threshold, because we need to be able to read below it.

I can only think of one case where this would be problematic: when there's one range tombstone below the GC threshold, one range tombstone above the GC threshold, and no point keys at all between or above them. In that case, the upper range tombstone would prevent use of a clearrange. But the same is true for any point tombstone above the GC threshold, and seems rare enough that it doesn't matter.

In the case where we're clearing a long run of versions for a single key, a range tombstone above it wouldn't prevent clearrange because the range key span would have zero size: `[startKey.Next(), startKey.Next())`.

Is there some other case I'm missing here?

Reply:
I guess the same holds if you have a lot of contiguous point tombstones that fall below the GC threshold under a live range tombstone. Is that a common scenario?

Reply:
Having read the actual GC batching code, I understand why this complicates things. That probably says something about the structure of the GC code, but I'm fine with leaving this to only GC point keys. `GCClearSubRangeKey` seems like a confusing name in that case.

Reply:
I was hoping to generalize this, but I suppose it's ok given that we're at least moving in the direction of a generalized `ClearRange` mechanism. However, it now has two very different and very specialized modes of operation, and I'm worried that this will be confusing enough that someone is going to use them in a wrong way and mess something up. We need to be extra, extra clear on the semantics here. We should also rename it to `GCClearRange`. I'd rewrite the comment to something like this:

>GCClearRange specifies a span that will be GCed using a Pebble range tombstone. This is *not* a generalized operation, it currently only has two very specialized modes:
>
> 1. Empty StartKeyTimestamp: used to delete an entire Raft range, both point keys and range keys. StartKey and EndKey must exactly match the Raft range bounds, and all data (including point tombstones) must be covered by an MVCC range tombstone, otherwise an error is returned. This is intended for use with SQL schema GC, which deletes e.g. an entire table using a single MVCC range tombstone.
>
> 2. Non-empty StartKeyTimestamp: deletes only point key versions from the given StartKey and StartKeyTimestamp, leaving newer versions of StartKey and all MVCC range keys intact. This is intended for bulk GC of long runs of point key garbage, with range keys GCed separately.
>
> In both cases, the request will error if any data is found above the GC threshold.
>
> TODO: generalize this to clear both point keys and range keys in an arbitrary key span (with or without a start key timestamp). The GC code currently isn't structured in a way that can easily process point keys and range keys simultaneously.